### DISCUSSION TITLE
PR for deferrable relation support in schema

### DISCUSSION THREAD

**Discussion by yume-chan** [prisma/prisma#14064] [Q&A]
Recently I encountered the same issue as #3502, where I have two models both reference each other, so it's not possible to create them normally. However, PostgreSQL has this deferrable constraint feature, which when opt-in, defers the check to transaction commit, so I can use an invalid value to create first row, then create the other side with correct id, and finally update the first row with correct id.

```
model A {
  id Int @id @default(autoincrement())

  b   B   @relation(fields: [bId], references: [id])
  bId Int
}

model B {
  id Int @id @default(autoincrement())

  a   A   @relation(fields: [aId], references: [id])
  aId Int
}
```

This feature requires `DEFERRABLE` and `INITIALLY DEFERRED` attributes on the constraint, so it needs support in Prisma schema. Previously I edited the migration file manually, but I don't want to do that in production.

So, I did some quick experiment on my branch https://github.com/yume-chan/prisma-engines/tree/deferrable-relation. I extended the `@relation` attribute with extra `deferrable: bool` and `initiallyDeferred: bool` arguments. In my small manual test, it can work.

```
model A {
  id Int @id @default(autoincrement())

  b   B   @relation(fields: [bId], references: [id], deferrable: true, initiallyDeferred: true)
  bId Int
}

model B {
  id Int @id @default(autoincrement())

  a   A   @relation(fields: [aId], references: [id], deferrable: true, initiallyDeferred: true)
  aId Int
}
```

I have read https://github.com/prisma/prisma-engines/blob/main/CONTRIBUTING.md so I'm asking here, whether Prisma team will accept a PR for this feature? Of course, I also need help on code reviewing, writing tests/docs, and updating the language server.

**Comment by yume-chan** [comment]
Forgot to mention, SQLite also supports deferrable foreign key (but not other constraints): https://www.sqlite.org/foreignkeys.html#fk_deferred. If a PR will be accepted, I can also add SQLite support if time allows.

**Comment by janpio** [comment]
That is a very difficult to answer without doing the full research into this feature. This would need to work in Migrations, but also Introspection to be picked up correctly at least.

To summarize your proposal:

1. This functionality exists in PostgreSQL and SQLite 
	- https://www.postgresql.org/docs/current/sql-set-constraints.html
	- https://www.sqlite.org/foreignkeys.html#fk_deferred
2. API for it should be `deferrable: bool` and `initiallyDeferred: bool` on `@relation`

Correct?

If you confirm, I can raise this with the team responsible for such features and see if they what they think.

**Reply by yume-chan** [reply]
Doc for Postgres show be in https://www.postgresql.org/docs/14/sql-createtable.html

Postgres supports many deferrable constraint related features:

* Deferrable foreign key, unique constraint, and primary key: Today it's not possible to declare them in Prisma schema. My proposal (adding `deferrable: bool` and `initiallyDeferred: bool` to `@relation`) adds support only for foreign keys, but this pattern can also be added to `@@unique` and `@id` in the future.
* Set whether a constraint check is deferred in `START TRANSACTION`, `SET TRANSACTION` and `SET CONSTRAINTS`: There features need to be added to the Prisma client. However, `initiallyDeferred` is a good default value, and they can already be invoked using `$executeRaw`, so this part can be in low priority.

SQLite only supports deferrable foreign keys.

My experimental implementation already supports both migration and introspection for Postgres.

**Comment by Jolg42** [comment]
Related https://github.com/prisma/prisma/issues/8807

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
